src/components/NotesMindMap/NotesMindMapReducer.js

```diff
@@ -49,59 +48,65 @@ export const notesMindMapReducer = (state = defaultState, { type, data }) => {
 
   const handleNoteNameUpdateRequestSuccessAction = () => {
     const updatedNote = data;
-    const newState = cloneTreeInState();
-    const noteToUpdate = tree(newState.rootNote).find(note => note.id === updatedNote.id);
-    noteToUpdate.name = updatedNote.name;
-    return newState;
+    let notes = [ ...state.notes ]
+    notes = notes.filter(note => note.id !== updatedNote.id)
+    notes.push({ id: updatedNote.id, label: updatedNote.name, name: updatedNote.name, isNote: updatedNote.isNote });
+    return { ...state, notes };
   };
 
   const handleCreateNoteSuccessAction = () => {
     const newNote = data;
-    const newState = cloneTreeInState();
-    const parentNote = tree(newState.rootNote).find(note => newNote.parent.id === note.id);
-    parentNote.children.push(newNote);
-    return newState;
+    let notes = [ ...state.notes ]
+    let edges = [ ...state.edges ]
+    notes.push({ id: newNote.id, name: newNote.name, label: newNote.name, isNote: newNote.isNote, group: 'children' });
+    edges.push({ from: newNote.parent.id, to: newNote.id });
+    notes = addGroupTagToNote(notes, state.selectedNote.id)
+      return { ...state, notes, edges };
   };
 
   const handleDeleteNoteRequestSuccessAction = () => {
     const noteToDelete = data;
-    const newState = cloneTreeInState();
-    newState.selectedNote = noteToDelete.parent
-    noteToDelete.parent.children = noteToDelete.parent.children.filter(
-      child => child.id !== noteToDelete.id
-    );
-    return newState;
+    let notes = [ ...state.notes ]
+    let edges = [ ...state.edges ]
+    let graph = removeNoteFromGraph(notes, edges, noteToDelete)
+    notes = graph.notes
+    edges = graph.edges
```

Comment:
Looks a bit weird... Can't you make the `removeNoteFromGraph` function to be pure, so you don't have to clone the arrays before using it? So we could have
```
const { notes, edges } = removeNoteFromGraph(notes, edges, noteToDelete);
return { ...state, notes, edges };
```
Or even
```
return { ...state, ...removeNoteFromGraph(notes, edges, noteToDelete) };
```
?